recipes-ni/ni-base-system-image-tests/files/fs_permissions_known.py

```diff
@@ -27,13 +27,161 @@ def known_permissions_tree():
                 'bzImage': system_link(f'bzImage-{bash("$(uname -r)")[0]}'),
             }
         },
+        'etc': {
+            'fstab': permissions(0o0644, 'lvuser', 'ni', FT_REG)
+        },
+        'home': {
+            '.': system_dir,
+            '*': system_dir,
+            'lvuser': permissions(0o2775, 'lvuser', 'ni', FT_DIR),
+            'root': system_link('/home/admin'),
+            'webserv': permissions(0o2755, 'webserv', 'ni', FT_DIR)
+        },
         'lib': {
             '.': system_dir,
             'modules': {
                 '.': system_dir,
                 '$(uname -r)': {
                     '.': system_dir,
                     '**': system_hier,
+                    'build': {
```

Comment:
I have some concern that listing all the executables and symlinks explicitly could be brittle. If I understand correctly, any time an executable or symlink is added or removed under one of these directories, it is going to cause a test failure and require modifying this code.

For symlinks, it seems better to only validate the permissions (always 0777, correct?) and ownership and just ignore the target of the link. I think our intention in testing known permissions is primarily to ensure secure/appropriate permissions/ownership, which I think can be done without examining the link target. I think we could do this pretty easily by extending system_hier to only check permissions/ownership of links, similar to how it handles both files/directories now. Then we could remove all the symlink entries from this tree.

Regarding executables, I think we can deal with many of them using file extensions (.sh, .pl) as you are now. The trick is the compiled binaries. We could possibly do this dynamically by using some python equivalent of the 'file' shell command or examining ELF headers or similar. I thought we touched on that idea previously. Did we decide it wasn't feasible?

Reply:
I don't remember discussing that idea, but I'll see what I can do. I don't like listing them all out either.

Reply:
If `readelf -h <file-name>`  succeeds then a file is a program. It will fail on symlinks, directories, and files that are not ELF programs. There might be a way to feed that into find.

Reply:
My suggestion previously with `subset_permissions` takes an arbitrary predicate to decide what a particular file should look like. That's one option here: make a predicate that runs `readelf -h` or checks file extensions, and then dispatch appropriately.

Reply:
As Mike suggested, I was thinking of a predicate that would dynamically examine each file already placed in the tree via basic wildcard expansion.

That being said, I don't think **readelf -h** will identify nearly all the files that are currently marked as executable that do not have a file extension. Most of them look to be scripts of some kind. This isn't perfect, but it's closer:
**file <file_path> | grep executable**

It's already clear both extension-based rules and this mechanism are going to find some files that "look" executable but don't actually have execute permissions when installed (and vice-versa in the case of build/scripts/atomic/fallbacks/*). Perhaps we should have a short discussion and determine exactly how to proceed...

Reply:
We might be able to use `grep -rIL`

Reply:
> For symlinks, it seems better to only validate the permissions (always 0777, correct?) and ownership and just ignore the target of the link. I think our intention in testing known permissions is primarily to ensure secure/appropriate permissions/ownership, which I think can be done without examining the link target. I think we could do this pretty easily by extending system_hier to only check permissions/ownership of links, similar to how it handles both files/directories now. Then we could remove all the symlink entries from this tree.

I don't think you resolved this bit of feedback.

Reply:
Ah, yes. I missed that. I'll add them too